### ISSUE TITLE
Read blob without downloaded it

### ISSUE THREAD

**Issue by FrancoisMasson1990** [googleapis/python-storage#124]
is it possible to read blobs without downloading them in google cloud ?

**Comment by FrancoisMasson1990** [comment]
> Hi @FrancoisMasson1990 - I assume you're using `google-cloud-storage`? Could you elaborate on what you're trying to do, or provide us with a code snippet?
> 
> There are `download_as_string` and `download_to_file` methods for [Blobs](https://googleapis.dev/python/storage/latest/blobs.html#google.cloud.storage.blob.Blob.download_as_string).
> 
> Some other services do allow you to provide data in blobs by giving URIs. `gs://...`
> 
> For example, in AutoML you are expected to upload your data to a storage bucket. [doumentation](https://cloud.google.com/vision/automl/docs/prepare#csv)

**Comment by FrancoisMasson1990** [comment]
Hello @busunkim96 

Indeed I am using google-cloud-storage. In a bucket I have a folder where some files are saved in a .db format (SQLite 3 format). What I would like to do is to read the data inside these files without download it locally. "data = blob.read_data" in a pythonic way
Right now, I access the bucket, download the files then read them locally.
Is it possible ?

**Comment by busunkim96** [comment]
I think `download_as_string` will work for your use case:

```python
from google.cloud import storage


def download_blob(bucket_name, source_blob_name):
    """Downloads a blob from the bucket."""
    # bucket_name = "your-bucket-name"
    # source_blob_name = "storage-object-name"

    storage_client = storage.Client()

    bucket = storage_client.bucket(bucket_name)
    blob = bucket.blob(source_blob_name)
    content = blob.download_as_string()

    return content
```

**Comment by FrancoisMasson1990** [comment]
Hello @busunkim96,

it returns bytes object. Do you know you to convert it to the original db format ?

**Comment by frankyn** [comment]
Hi @FrancoisMasson1990,

IIUC, you're attempting to load the data stored in Cloud Storage (a sqlite database) into memory.

Based on the stackoverflow question here: https://stackoverflow.com/questions/3850022/how-to-load-existing-db-file-to-memory-in-python-sqlite3, you'd need to convert your db format into sql statements that would then be stored in Cloud Storage as an object and then execute sql script by downloading into memory.

Convert file.db into sql commands:

```python
import sqlite3
from io import StringIO

def sql_dump(bucket_name, dest_blob_name):
    # Read database to tempfile
    con = sqlite3.connect("path/to/file.db")
    temp = StringIO()
    for line in con.iterdump():
        temp.write('%s\n' % line)
    con.close()
    temp.seek(0)
   # you would store this in Cloud Storage to load later.
   storage_client = storage.Client()

    bucket = storage_client.bucket(bucket_name)
    blob = bucket.blob(dest_blob_name)
    blob.upload_from_string(temp)
```

You'd do something similar to the following once the database is represented as sql commands:

```python
import sqlite3
from google.cloud import storage

def load_sqlite_db_in_memory(bucket_name, source_blob_name):
     storage_client = storage.Client()

    bucket = storage_client.bucket(bucket_name)
    blob = bucket.blob(source_blob_name)
    content = blob.download_as_string()

    # Create a database in memory
    sqlite = sqlite3.connect(":memory:")
    sqlite.cursor().executescript(content)
    sqlite.commit()
  
    return sqlite
```

**Comment by frankyn** [comment]
Hi @FrancoisMasson1990, did this help unblock you? If so I'd like to close out this issue.

**Comment by sonsharma** [comment]
Hello,

I'm trying to read a shapefile placed in one of my buckets in gcs but I'm unable to. I don't want to download the file in my local while reading or use geopandas to read them. I'm running everything on my local and I have all the accesses to gcp bucket. Below is my code:
```
file_path = "gs://my-bucket/country/shapefiles/my-shapefile.shp"
sf_shape = shapefile.Reader(file_path)
```

I'm able to read the csv files in this manner but I can't read shapfiles directly from gcs. Also, I tried the following:
```
fs = gcsfs.GCSFileSystem(project="citi-ventures")
with fs.open(file_path, 'w') as f:
     shapefile.Reader(f.read())`

```
file_path = "country/shapefiles/my-shapefile.shp"
storage_client = storage.Client("project-name")
blob = storage_client.get_bucket(bucket_name).blob(file_path)
shapefile.Reader(blob.download_to_bytes())
```

Please help!! Thanks in advance

**Comment by frankyn** [comment]
Hi @sonsharma could you please reopen this as a new issue so we can track this separately?

**Comment by sonsharma** [comment]
> Hi @sonsharma could you please reopen this as a new issue so we can track this separately?

This is the issue: https://github.com/googleapis/python-storage/issues/764

Thank you!!
